astropy/table/table.py

```diff
@@ -372,6 +381,139 @@ def filled(self, fill_value=None):
             data = self
         return self.__class__(data, meta=deepcopy(self.meta))
 
+    @property
+    def indices(self):
+        '''
+        Returns the indices associated with columns of the table.
+        '''
+        lst = []
+        for column in self.columns.values():
+            for index in self._get_indices(column):
+                if sum([index is x for x in lst]) == 0: # ensure uniqueness
+                    lst.append(index)
+        return lst
+
+    def _get_indices(self, col):
+        '''
+        Returns the indices associated with col, or [] if col
+        is a mixin rather than a regular column.
+        '''
+        return getattr(col, 'indices', [])        
+
+    def add_index(self, colnames, impl=None):
+        '''
+        Inserts a new index among one or more columns. The input parameter
+        may be either a list of column names or a single column name.
+        '''
+        if isinstance(colnames, six.string_types):
+            colnames = (colnames,)
+        columns = self.columns[tuple(colnames)].values()
+        index = Index(columns, impl=impl)
+        for col in columns:
+            col.add_index(index)
+
+    def remove_indices(self, colname):
+        '''
+        Remove all indices involving the given column.
+        '''
+        col = self.columns[colname]
+        for index in self.indices:
+            try:
+                index.col_position(col)
+            except ValueError:
+                pass
+            else:
+                for c in index.columns:
+                    c.indices.remove(index)
+
+    def where(self, expression, *vals):
+        '''
+        Returns all rows matching the given criteria.
+        Language syntax in Backus-Naur Form:
+
+        <query> := "[" <column name> "]" <criterion> | <query> "and" <query>
+        <criterion> := "in" <range> | "=" <value>
+        <range> := <left-paren> <value> "," <value> <right-paren>
+        <value> := "{" <number> "}"
+        <left-paren> := "[" | "("
+        <right-paren> := "]" | ")"
+
+        Example: table.where('[col0] in [{0}, {1}) and [col1] = {2}', 1, 5, 3)
```

Comment:
I'd like to see some doctests here showing examples of usage.  It's a little unclear from the BNF grammar what everything does.

Also, this, like all docstrings, should be in numpydoc format.